llvm/lib/Target/AMDGPU/SIFixSGPRCopies.cpp

```diff
@@ -901,6 +911,34 @@ bool SIFixSGPRCopies::lowerSpecialCase(MachineInstr &MI,
                                       MI, MI.getDebugLoc())) {
       I = std::next(I);
       MI.eraseFromParent();
+    } else if (SrcReg.isVirtual() && TRI->getRegSizeInBits(SrcReg, *MRI) ==
+                                         TRI->getRegSizeInBits(DstReg, *MRI)) {
+      auto I = MI.getIterator();
+      auto E = MI.getParent()->end();
+      // COPY can be erased if all its uses can be converted to waterfall.
+      bool CanErase = true;
+      // Only search current block since phyreg's def & use cannot cross
+      // blocks when MF.NoPhi = false.
+      while (++I != E) {
+        // Currently, we only support waterfall on SI_CALL_ISEL.
+        if (I->getOpcode() == AMDGPU::SI_CALL_ISEL) {
+          MachineInstr *UseMI = &*I;
+          for (unsigned i = 0; i < UseMI->getNumOperands(); ++i) {
+            if (UseMI->getOperand(i).isReg() &&
+                UseMI->getOperand(i).getReg() == DstReg) {
+              MachineOperand *MO = &UseMI->getOperand(i);
+              MO->setReg(SrcReg);
+              WaterFalls[UseMI].MOs.push_back(MO);
+              WaterFalls[UseMI].SGPRs.push_back(DstReg);
```

Comment:
You can do this with a single lookup on the map.

```suggestion
              V2PysSCopyInfo &V2SCopyInfo = WaterFalls[UseMI];
              V2SCopyInfo.MOs.push_back(MO);
              V2SCopyInfo.SGPRs.push_back(DstReg);
```